src/hardware/driver/output.rs

```diff
@@ -0,0 +1,120 @@
+        }
+    }
+}
+
+pub mod sm {
+    use super::*;
+    statemachine! {
+        transitions: {
+            *Disabled + Enable = EnableWaitRelay,
+            EnableWaitRelay + RelayDone = RampCurrent,
+            RampCurrent + CurrentStep / increment_current = RampCurrent,
+            RampCurrent + CurrentFinal = Enabled,
+            Enabled + Disable / reset_iir = DisableWaitRelay,
+            DisableWaitRelay + RelayDone  = Disabled,
+            Disabled + Disable = Disabled,
+            Enabled + Enable = Enabled
+        }
+    }
+}
+
+impl sm::StateMachineContext for Output {
+    fn increment_current(&mut self) {
+        self.ramp_iir.y_offset += Output::RAMP_STEP;
+    }
+
+    fn reset_iir(&mut self) {
+        self.ramp_iir.y_offset = 0.;
+    }
+}
+
+#[derive(Debug)]
+pub enum Error {
+    Output(sm::Error),
+    Relay(relay::sm::Error),
+}
+
+impl From<sm::Error> for Error {
+    fn from(err: sm::Error) -> Error {
+        Error::Output(err)
+    }
+}
+
+impl From<relay::sm::Error> for Error {
+    fn from(err: relay::sm::Error) -> Error {
+        Error::Relay(err)
+    }
+}
+
+impl sm::StateMachine<Output> {
+    /// Start enabling sequence. Returns `relay delay` or an error is a SM is in an illegal state.
+    pub fn enable(
+        &mut self,
+        relay: &mut relay::sm::StateMachine<relay::Relay<I2c1Proxy>>,
+    ) -> Result<fugit::MillisDuration<u64>, Error> {
+        self.process_event(sm::Events::Enable)?;
+        relay.enable().map_err(|err| err.into())
+    }
+
+    /// Start disabling sequence. Returns `relay delay` or an error is a SM is in an illegal state.
+    pub fn disable(
+        &mut self,
+        relay: &mut relay::sm::StateMachine<relay::Relay<I2c1Proxy>>,
+    ) -> Result<fugit::MillisDuration<u64>, Error> {
+        self.process_event(sm::Events::Disable)?;
+        relay.disable().map_err(|err| err.into())
+    }
+
+    /// Handle realays done. Returns `Some(ramp delay)` if output is enabling and `None` if it is disabling.
+    pub fn relay_done(&mut self) -> Option<fugit::MillisDuration<u64>> {
+        self.process_event(sm::Events::RelayDone).unwrap();
+        if *self.state() == sm::States::RampCurrent {
+            Some(Output::RAMP_DELAY)
+        } else {
+            None
+        }
+    }
+
+    /// Handle current ramp step. Returns `Some(ramp delay)` unless the ramp is done.
+    /// In this case returns `None`.
+    pub fn handle_ramp(
+        &mut self,
+        iir: iir::IIR<f32>,
+    ) -> Option<fugit::MillisDuration<u64>> {
+        if self.context().ramp_iir.y_offset >= iir.y_offset {
+            self.process_event(sm::Events::CurrentFinal).unwrap();
+            None
+        } else {
+            self.process_event(sm::Events::CurrentStep).unwrap();
+            Some(Output::RAMP_DELAY)
+        }
+    }
+
+    pub fn iir(&mut self) -> iir::IIR<f32> {
```

Comment:
Better give back a reference here. This will always copy. Instead if you let the user do the copy (`let iir = *state_machine.iir()`, it's a bit more flexible.